allocator_test.go

```diff
@@ -1,23 +1,239 @@
 package allocator_test
 
 import (
-	"testing"
-
 	"github.com/irfansharif/allocator"
 	"github.com/stretchr/testify/require"
+	"math/rand"
+	"testing"
 )
 
-func TestAllocator(t *testing.T) {
-	var rs []allocator.Range
-	for i := 0; i < 3; i++ {
-		rs = append(rs, allocator.NewRange(allocator.RangeID(i), 3, nil, nil))
+func TestGIVENReplicationConstraintTHENAppropriateReplicasAllocated(t *testing.T) {
+	const RangeSizeLimit = 20
+	const ReplicationFactor = 3
+	const TestClusterSize = 64
+	builtCluster := buildCluster(TestClusterSize, nodeCapacitySupplier(TestClusterSize, 0, 1))
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(allocator.RangeID(initIndex), ReplicationFactor, nil, nil)
+	}
+	for rangeIndex := range rangesToAllocate {
+		underTest := allocator.New(rangesToAllocate[:rangeIndex], builtCluster)
+		status, allocation := underTest.Allocate()
+		require.True(t, status)
+		for _, nodeAssignments := range allocation {
+			require.Equal(t, len(nodeAssignments), ReplicationFactor)
+			require.True(t, isValidNodeAssignment(nodeAssignments, TestClusterSize))
+			require.True(t, isEachReplicaAssignedToDifferentNode(nodeAssignments, TestClusterSize))
+		}
+	}
+}
+
+func TestGIVENReplicationConstraintWHENInsufficientNodesTHENAllocationFails(t *testing.T) {
+	const RangeSizeLimit = 20
```

Comment:
What's RangeSizeLimit? Do you mean numRanges? Also, use numNodes instead of TestClusterSize -- it's clearer. Don't capitalize replicationFactor.